restapi/src/test/java/io/stargate/web/docsapi/service/DocsShredderTest.java

```diff
@@ -0,0 +1,1237 @@
+    assertThat(arraysMatch(bindParams.get(3), expectedRows.get(3))).isTrue();
+    assertThat(arraysMatch(bindParams.get(4), expectedRows.get(4))).isTrue();
+
+    Set<String> firstLevelKeyset = new HashSet<>(results.right);
+    assertThat(firstLevelKeyset.size()).isEqualTo(3);
+    assertThat(firstLevelKeyset).contains("some", "array", "nested");
+  }
+
+  @Test
+  public void shredJson_happyPathEmptyObject() {
+    List<Object[]> expectedRows = getExpectedEmptyObjectRows();
+    when(db.newBindMap(anyList())).thenCallRealMethod();
+    when(db.treatBooleansAsNumeric()).thenReturn(false);
+    String payload = "{\"some\":{}, \"array\":[[], true, 3], \"nested\": {\"structure\": 23.456}}";
+    ImmutablePair<List<Object[]>, List<String>> results =
+        service.shredJson(surfer, db, Collections.emptyList(), "documentId", payload, false);
+    List<Object[]> bindParams = results.left;
+    assertThat(bindParams.size()).isEqualTo(expectedRows.size());
+    assertThat(arraysMatch(bindParams.get(0), expectedRows.get(0))).isTrue();
+    assertThat(arraysMatch(bindParams.get(1), expectedRows.get(1))).isTrue();
+    assertThat(arraysMatch(bindParams.get(2), expectedRows.get(2))).isTrue();
+    assertThat(arraysMatch(bindParams.get(3), expectedRows.get(3))).isTrue();
+    assertThat(arraysMatch(bindParams.get(4), expectedRows.get(4))).isTrue();
+
+    Set<String> firstLevelKeyset = new HashSet<>(results.right);
+    assertThat(firstLevelKeyset.size()).isEqualTo(3);
+    assertThat(firstLevelKeyset).contains("some", "array", "nested");
+  }
+
+  @Test
+  public void shredJson_emptyObject() {
+    String payload = "{}";
+    ImmutablePair<List<Object[]>, List<String>> results =
+        service.shredJson(surfer, db, Collections.emptyList(), "documentId", payload, false);
+    List<Object[]> bindParams = results.left;
+    assertThat(bindParams.size()).isEqualTo(0);
+
+    Set<String> firstLevelKeyset = new HashSet<>(results.right);
+    assertThat(firstLevelKeyset.size()).isEqualTo(0);
+  }
+
+  @Test
+  public void shredJson_errorPatching() {
+    String payload = "[1, true, 3]";
+    assertThatThrownBy(
+            () ->
+                service.shredJson(surfer, db, Collections.emptyList(), "documentId", payload, true))
+        .hasMessage("A patch operation must be done with a JSON object, not an array.");
+  }
+
+  @Test
+  public void shredJson_errorMaxDepth() {
+    String payload =
+        "[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[\"tooDeep\"]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]";
+    assertThatThrownBy(
+            () ->
+                service.shredJson(
+                    surfer, db, Collections.emptyList(), "documentId", payload, false))
+        .hasMessage("Max depth of 64 exceeded.");
+  }
+
+  private List<Object[]> getExpectedRows() {
+    Object[] expectedRow0 =
+        new Object[] {
+          "documentId",
+          "some",
+          "",
```

Comment:
Might be more readable by constructing programmatically, given our Formatter insisting on one element per line...
That is, allocate array, fill with ""s, override ones that are not empty?